src/jwt/mod.rs

```diff
@@ -20,11 +20,98 @@ new_type![
     JsonWebTokenContentType(String)
 ];
 
+#[derive(Error, Debug)]
+pub enum JsonWebTokenTypeNormalizeError {
+    #[error("Has parameters set without application specifies like JWT;param=true (should be application/JWT;param=true)")]
+    ParameterWithoutApplication(),
+
+    #[error(
+        "The web token type is complex (contains /) but does not start with an application type."
+    )]
+    ComplexTypeWithoutApplication(),
+}
+
 new_type![
+    /// JSON web token type field (typ)
+    ///
+    /// This types stores the raw (deserialized) value.
+    ///
+    /// To compare two different JSON Web Token types, please use the normalized version via [`JsonWebTokenType::normalize`]
     #[derive(Deserialize, Hash, Ord, PartialOrd, Serialize)]
     JsonWebTokenType(String)
+
+    impl {
+        /// Expands a [`JsonWebTokenType`] and produces a [`NormalizedJsonWebTokenType`] according to RFC2045 and RFC7515.
+        ///
+        /// See [RFC 2045 section 5.1](https://tools.ietf.org/html/rfc2045#section-5.1) for the full Content-Type Header Field spec.
+        /// See [RFC 7515 section 4.19](https://tools.ietf.org/html/rfc7515#section-4.1.9) for specific requirements of JSON Web Token Types.
+        pub fn normalize(&self) -> Result<NormalizedJsonWebTokenType, JsonWebTokenTypeNormalizeError> {
+            self.try_into()
+        }
+    }
+];
+
+new_type![
+    /// JSON web token type field (typ)
+    ///
+    /// This types stores the raw (deserialized) value.
+    /// To retrieve a normalized value according to RFC2045 and RFC7515 see [`JsonWebTokenType::normalize`]
+    ///
+    /// See [RFC 2045 section 5.1](https://tools.ietf.org/html/rfc2045#section-5.1) for the full Content-Type Header Field spec.
+    /// See [RFC 7515 section 4.1.9](https://tools.ietf.org/html/rfc7515#section-4.1.9) for specific requirements of JSON Web Token Types.
+    ///
+    /// It is recommended to only instantiate `NormalizedJsonWebTokenType` objects via [`JsonWebTokenType`] and then call [`JsonWebTokenType::normalize`].
+    ///
+    /// ```rust
+    /// # use openidconnect::{NormalizedJsonWebTokenType, JsonWebTokenType};
+    /// let token_type: NormalizedJsonWebTokenType = JsonWebTokenType::new("jwt+at".to_string())
+    ///     .normalize()
+    ///     .unwrap();
+    ///
+    /// // normalized value looks like this
+    /// assert_eq!(*token_type, "application/jwt+at")
+    /// ```
+    ///
+    /// When you really want to create this type directly, please check the following items:
```

Comment:
this really suggests that we shouldn't allow direct construction (without `normalize` or `TryFrom`), which means not using `new_type!`. it shouldn't be too bad to implement this type manually though (without the `new` function): https://github.com/ramosbugs/openidconnect-rs/blob/fd404985ef6c8e546f951191f4e1bc791615f5ca/src/macros.rs#L122-L146